src/parser/pattern.lisp

```diff
@@ -0,0 +1,120 @@
+(defpackage #:coalton-impl/parser/pattern
+  (:use
+   #:cl
+   #:coalton-impl/parser/base)
+  (:shadowing-import-from
+   #:coalton-impl/parser/base
+   #:parse-error)
+  (:local-nicknames
+   (#:cst #:concrete-syntax-tree)
+   (#:util #:coalton-impl/util))
+  (:export
+   #:pattern                            ; STRUCT
+   #:pattern-source                     ; ACCESSOR
+   #:pattern-list                       ; TYPE
+   #:pattern-var                        ; STRUCT
+   #:make-pattern-var                   ; ACCESSOR
+   #:pattern-var-name                   ; ACCESSOR
+   #:pattern-literal                    ; STRUCT
+   #:make-pattern-literal               ; CONSTRUCTOR
+   #:pattern-literal-value              ; ACCESSOR
+   #:pattern-wildcard                   ; STRUCT
+   #:make-pattern-wildcard              ; ACCESSOR
+   #:pattern-constructor                ; STRUCT
+   #:make-pattern-constructor           ; CONSTRUCTOR
+   #:pattern-constructor-name           ; ACCESSOR
+   #:pattern-constructor-patterns       ; ACCESSOR
+   #:parse-pattern                      ; FUNCTION
+   ))
+
+(in-package #:coalton-impl/parser/pattern)
+
+(defstruct (pattern
+            (:constructor nil)
+            (:copier nil))
+  (source (util:required 'source) :type cons :read-only t))
+
+(defun pattern-list-p (x)
+  (and (alexandria:proper-list-p x)
+       (every #'pattern-p x)))
+
+(deftype pattern-list ()
+  '(satisfies pattern-list-p))
+
+(defstruct (pattern-var
+            (:include pattern)
+            (:copier nil))
+  (name (util:required 'name) :type identifier :read-only t))
+
+(defstruct (pattern-literal
+            (:include pattern)
+            (:copier nil))
+  (value (util:required 'value) :type util:literal-value :read-only t))
+
+(defstruct (pattern-wildcard
+            (:include pattern)
+            (:copier nil)))
+
+(defstruct (pattern-constructor
+            (:include pattern)
+            (:copier nil))
+  (name     (util:required 'name)     :type identifier   :read-only t)
+  (patterns (util:required 'patterns) :type pattern-list :read-only t))
+
+(defun parse-pattern (form file)
```

Comment:
In reading this code, maybe which is just my personal history as a lisper, I'm getting a little mentally exhausted at the "early return"/"nested when"/"declare now set later" style of coding. I think this style of coding is a welcome sign to complex bugs, since it invites extension simply by splicing in new code in the middle of existing code. `PROGN` is the root of all evil, etc.

While I sometimes think sticking to a strict "expression-only" style of coding can lead to obtuse code (especially when you need to do a complex initialization), in this code, it's getting hard to follow what invariants are true when, since what an expression means now depends on following a sequence of statements, as opposed to reading a nested structure.